hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/Histogram1D.java

```diff
@@ -8,61 +8,73 @@
+     * Creates the histogram explicitly and in full. Should be called at most once.
      */
-    private void createHistogram(final IColumn column, final IMembershipSet membershipSet
+    public void createHistogram(final IColumn column, final IMembershipSet membershipSet
                                 ,final IStringConverter converter ) {
+        if (this.buckets[0] != null) //a histogram had already been created
+            throw new IllegalAccessError("A histogram cannot be created twice");
+        for (int i = 0; i < this.bucketDescription.getNumOfBuckets(); i++)
+            this.buckets[i] = new Bucket1D();
         final IRowIterator myIter = membershipSet.getIterator();
-        int currIndex = myIter.getNextRow();
-        while (currIndex >= 0) {
-            if (column.isMissing(currIndex))
+        int currRow = myIter.getNextRow();
+        while (currRow >= 0) {
+            if (column.isMissing(currRow))
                 this.missingData++;
-            else
-                placeInBucket(column.asDouble(currIndex, converter));
-            currIndex = myIter.getNextRow();
+            else {
+                double val = column.asDouble(currRow,converter);
+                int index = this.bucketDescription.indexOf(val);
+                if (index >= 0)
+                    this.buckets[index].add(val,column.getObject(currRow));
+                else this.outOfRange++;
+            }
+            currRow = myIter.getNextRow();
         }
     }
 
-    private void placeInBucket(final double entry) {
-        final int index = this.bucketDescription.indexOf(entry);
-        if (index >= 0)
-            this.buckets[index].add(entry);
-        else this.outOfRange++;
+    public long getMissingData() { return this.missingData; }
+
+    public long getOutOfRange() { return this.outOfRange; }
+
+    /**
+     * @return the index's bucket or null if not been initialized yet
+     */
+    public Bucket1D getBucket(final int index) {
+        if ((index < 0) || (index >= this.bucketDescription.getNumOfBuckets()))
+            throw new IllegalArgumentException("bucket index out of range");
+        return this.buckets[index];
+    }
+
+    /**
+     * @param  otherHistogram with the same bucketDescription
+     * @return a new Histogram which is the union of this and otherHistogram
+     */
+    public Histogram1D union( @NonNull Histogram1D otherHistogram) {
+        if (!this.bucketDescription.equals(otherHistogram.bucketDescription))
+            throw new IllegalArgumentException("Histogram union without matching buckets");
+        Histogram1D unionH = new Histogram1D(this.bucketDescription);
+        for (int i = 0; i < unionH.bucketDescription.getNumOfBuckets(); i++)
+            unionH.buckets[i] = this.buckets[i].union(otherHistogram.buckets[i]);
+        unionH.missingData = this.missingData + otherHistogram.missingData;
+        unionH.outOfRange = this.outOfRange + otherHistogram.outOfRange;
+        return unionH;
```

Comment:
You will have to think about addition of histograms on sampled data.
Probably the results have to be scaled up in order for addition to make sense.